src/VisualStudio/Core/Def/Options/PackageSettingsPersister.cs

```diff
@@ -40,9 +44,28 @@ private async Task InitializeAsync(CancellationToken cancellationToken)
         {
             await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
 
-            _lazyRoslynPackage = await RoslynPackage.GetOrLoadAsync(_threadingContext, _serviceProvider, cancellationToken).ConfigureAwait(true);
-            Assumes.Present(_lazyRoslynPackage);
+            // We attempt to get or load the RoslynPackage to proceed with the core initialization.
```

Comment:
This seems like a workaround this chick and egg problem; the package retrieval fails because you are overlapping with it.

Why don't you just reverse the dependency, have the package call into this "persister" (via a private interface) passing itself when its done?

Reply:
> Why don't you just reverse the dependency, have the package call into this "persister" (via a private interface) passing itself when its done?

I thought about it but there is no clear point in RoslynPackage where we know the package has finished loading. However, it seems the PackageSettingsPersister doesn't really care about the package loading stage, it just needs a reference to the instance of RoslynPackage to hookup an event handler. I'll implement your suggestion and confirm if all works, thanks!

Reply:
@davkean There's one bit I'm still not quite following here -- if an async package load is already underway, what is IVsShell7's LoadPackageAsync doing? I would have assumed it also waits until the package is done loading, but maybe not?